tools/protoc-gen-docs/htmlGenerator.go

```diff
@@ -0,0 +1,869 @@
+		if g.grouping {
+			g.print("<h2 id='Enumerations'>Enumerations</h2>")
+		}
+
+		for _, name := range enums {
+			e := enumMap[name]
+			g.currentFile = e.fileDesc()
+			g.generateEnum(e)
+		}
+	}
+
+	if len(messages) > 0 {
+		if g.grouping {
+			g.print("<h2 id='Messages'>Messages</h2>")
+		}
+
+		for _, name := range messages {
+			message := messageMap[name]
+			g.currentFile = message.fileDesc()
+			g.generateMessage(message)
+		}
+	}
+
+	if len(services) > 0 {
+		if g.grouping {
+			g.print("<h2 id='Services'>Services</h2>")
+		}
+
+		for _, name := range services {
+			service := serviceMap[name]
+			g.currentFile = service.fileDesc()
+			g.generateService(service)
+		}
+	}
+	g.generateFileFooter()
+
+	return plugin.CodeGeneratorResponse_File{
+		Name:    proto.String(pkg.name + ".pb.html"),
+		Content: proto.String(g.buffer.String()),
+	}
+}
+
+func (g *htmlGenerator) generateFileHeader(pkg *packageDescriptor) {
+	if g.mode == jekyllHTML {
+		g.print("---")
+
+		if pkg.title != "" {
+			g.print("title: ", pkg.title)
+		} else {
+			g.print("title: ", pkg.name)
+		}
+
+		if pkg.overview != "" {
+			g.print("overview: ", pkg.overview)
+		}
+
+		if pkg.location != "" {
+			g.print("location: ", pkg.location)
+		}
+
+		g.print("type: reference")
+		g.print("layout: docs")
+		g.print("---")
+		g.print("<!-- Generated by protoc-gen-docs -->")
+	} else if g.mode == htmlPage {
+		g.print("<!DOCTYPE html>")
+		g.print("<html itemscope itemtype='https://schema.org/WebPage'>")
+		g.print("<!-- Generated by protoc-gen-docs -->")
+		g.print("<head>")
+		g.print("<meta charset='utf-8'>")
+		g.print("<meta name='viewport' content='width=device-width, initial-scale=1, shrink-to-fit=no'>")
+
+		if pkg.title != "" {
+			g.print("<meta name='title' content='", pkg.title, "'>")
+			g.print("<meta name='og:title' content='", pkg.title, "'>")
+			g.print("<title>", pkg.title, "</title>")
+		}
+
+		if pkg.overview != "" {
+			g.print("<meta name='description' content='", pkg.overview, "'>")
+			g.print("<meta name='og:description' content='", pkg.overview, "'>")
+		}
+
+		g.print(htmlStyle)
+
+		g.print("</head>")
+		g.print("<body>")
+		if pkg.title != "" {
+			g.print("<h1>", pkg.title, "</h1>")
+		}
+	} else if g.mode == htmlFragment {
+		g.print("<!-- Generated by protoc-gen-docs -->")
+		if pkg.title != "" {
+			g.print("<h1>", pkg.title, "</h1>")
+		}
+	}
+
+	g.print()
+
+	count := 0
+	for _, loc := range pkg.loc {
+		if loc.LeadingComments != nil {
+			g.generateComment(loc, "")
+			count++
+		}
+	}
+
+	if count == 0 {
+		g.warn(nil, "no comment for package %s", pkg.name)
+	}
+}
+
+func (g *htmlGenerator) generateFileFooter() {
+	if g.mode == htmlPage {
+		g.print("</body>")
+		g.print("</html>")
+	}
+}
+
+func (g *htmlGenerator) generateSectionHeading(name string) {
+	if g.grouping {
+		g.print("<h3 id='", name, "'>", name, "</h3>")
+	} else {
+		g.print("<h2 id='", name, "'>", name, "</h2>")
+	}
+	g.print("<section>")
+	g.print()
```

Comment:
consider renaming print to printLn, these empty `print()` are unclear what they print.. or consider creatinga  new function printEmptyLn()

Reply:
I renamed print to emit, and I removed all the empty print() statement since they were adding semantically useless blank lines anyway.